### DISCUSSION TITLE
导入导出 nanoid 包，有问题 require() of ES modules is not supported.

### DISCUSSION THREAD

**Discussion by zhuzeyu22** [midwayjs/midway#3359] [Q&A] [answered]
### 🐛 Describe the bug（描述 bug）

导入导出 nanoid 包，有问题 require() of ES modules is not supported.

import { nanoid } from 'nanoid';

Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: D:\workspace\et_web_server2\node_modules\nanoid\index.js
require() of ES modules is not supported.
require() of D:\workspace\et_web_server2\node_modules\nanoid\index.js from D:\workspace\et_web_server2\src\service\ICD.service.ts is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from D:\workspace\et_web_server2\node_modules\nanoid\package.json.

    at Module._extensions..js (internal/modules/cjs/loader.js:1015:13)
    at Object.require.extensions.<computed> [as .js] (D:\workspace\et_web_server2\node_modules\ts-node\src\index.ts:1608:43)
    at Module.load (internal/modules/cjs/loader.js:863:32)
    at Function.Module._load (internal/modules/cjs/loader.js:708:14)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (D:\workspace\et_web_server2\src\service\ICD.service.ts:8:1)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Module.m._compile (D:\workspace\et_web_server2\node_modules\ts-node\src\index.ts:1618:23)
    at Module._extensions..js (internal/modules/cjs/loader.js:1027:10) {
  code: 'ERR_REQUIRE_ESM'
}
[ Midway ] Auto reload. [change] service\ICD.service.ts

### Midway Versions（Midway 版本）

npx: 3 安装成功，用时 4.969 秒
>> Start to check your midway component version...

✓ @midwayjs/faas-typings(not installed)
✓ @midwayjs/fc-starter(not installed)
✓ @midwayjs/serverless-http-parser(not installed)
✖ @midwayjs/async-hooks-context-manager(current: 3.12.3, allow: ["3.9.0"])
✓ @midwayjs/axios(not installed)
✓ @midwayjs/bootstrap(3.9.0)
✓ @midwayjs/bull-board(not installed)
✓ @midwayjs/bull(not installed)
✓ @midwayjs/cache(not installed)
✓ @midwayjs/captcha(not installed)
✓ @midwayjs/casbin-redis-adapter(not installed)
✓ @midwayjs/casbin-typeorm-adapter(not installed)
✓ @midwayjs/casbin(not installed)
✓ @midwayjs/code-dye(not installed)
✓ @midwayjs/consul(not installed)
✓ @midwayjs/core(3.9.0)
✓ @midwayjs/cos(not installed)
✓ @midwayjs/cross-domain(not installed)
✓ @midwayjs/decorator(3.9.0)
✓ @midwayjs/etcd(not installed)
✓ @midwayjs/express-session(not installed)
✓ @midwayjs/faas(not installed)
✓ @midwayjs/grpc(not installed)
✓ @midwayjs/http-proxy(not installed)
✖ @midwayjs/i18n(current: 3.12.7, allow: ["3.9.0","3.9.9"])
✓ @midwayjs/info(3.9.0)
✓ @midwayjs/jwt(not installed)
✓ @midwayjs/kafka(not installed)
✓ @midwayjs/mikro(not installed)
✓ @midwayjs/mock(3.9.0)
✓ @midwayjs/mongoose(not installed)
✓ @midwayjs/oss(not installed)
✓ @midwayjs/otel(not installed)
✓ @midwayjs/passport(not installed)
✓ @midwayjs/process-agent(not installed)
✓ @midwayjs/prometheus-socket-io(not installed)
✓ @midwayjs/prometheus(not installed)
✓ @midwayjs/rabbitmq(not installed)
✓ @midwayjs/redis(not installed)
✓ @midwayjs/security(not installed)
✓ @midwayjs/sequelize(not installed)
✖ @midwayjs/session(current: 3.12.3, allow: ["3.9.0"])
✓ @midwayjs/socketio(not installed)
✓ @midwayjs/static-file(not installed)
✖ @midwayjs/swagger(current: 3.12.7, allow: ["3.9.0","3.9.3","3.9.8","3.9.9"])
✓ @midwayjs/tablestore(not installed)
✓ @midwayjs/typegoose(not installed)
✓ @midwayjs/typeorm(not installed)
✓ @midwayjs/upload(not installed)
✓ @midwayjs/validate(3.9.0)
✖ @midwayjs/version(current: 3.12.7, allow: ["3.9.0","3.9.1","3.9.2","3.9.3","3.9.4","3.9.5","3.9.6","3.9.7","3.9.8","3.9.9"])
✓ @midwayjs/view-ejs(not installed)
✓ @midwayjs/view-nunjucks(not installed)
✓ @midwayjs/view(not installed)
✓ @midwayjs/express(not installed)
✓ @midwayjs/koa(3.9.0)
✓ @midwayjs/web(not installed)
✓ @midwayjs/ws(not installed)
**********************************************************************
>> Check complete, found  5  problem.
>> Use -u to show update list.
>> Use -u -w to write file.
>> Please check the result above.
**********************************************************************

**Comment by czy88840616** [comment, marked as answer]
非问题，cjs 和 esm 不兼容，以下方案二选一

1、使用仅支持 cjs 的包，比如老版本的 nanoid
2、使用 esm 版本的框架，https://midwayjs.org/docs/esm

**Reply by zhuzeyu22** [reply]
可能需要老版本，因为业务需要在 node 12 环境编译
